src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -431,12 +440,25 @@ void ReadFromRemote::addLazyPipe(Pipes & pipes, const ClusterProxy::SelectStream
     if (filter_actions_dag)
         pushed_down_filters = std::make_shared<const ActionsDAG>(filter_actions_dag->clone());
 
+    const StorageID resolved_id = context->resolveStorageID(shard.main_table ? shard.main_table : main_table);
+    const StoragePtr storage = DatabaseCatalog::instance().tryGetTable(resolved_id, context);
+    if (!storage)
+    {
+        throw Exception(ErrorCodes::UNKNOWN_TABLE, "Storage with id {} not found", resolved_id);
+    }
+
+    const auto * replicated_storage = dynamic_cast<const StorageReplicatedMergeTree *>(storage.get());
+    if (!replicated_storage)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected lazy remote read from a non-replicated table: {}", storage->getName());
```

Comment:
It is possible to trigger this in case of fail point, see src/Interpreters/ClusterProxy/SelectStreamFactory.cpp#L197

Reply:
That's a good catch. I changed it so that this cast will be executed only if `use_delayed_remote_source` is not used.